acasclient/protocol.py

```diff
@@ -0,0 +1,70 @@
+from acasclient.lsthing import (CodeValue, SimpleLsThing, LsThing, clob, ACAS_DDICT, ACAS_AUTHOR)
+
+from datetime import datetime
+
+# Constants
+ACAS_LSTHING = 'ACAS LsThing'
+META_DATA = 'protocol metadata'
+LS_TYPE = 'default'
+LS_KIND = 'default'
+PREFERRED_LABEL = 'protocol name'
+STATUS = 'protocol status'
+NOTEBOOK_PAGE = 'notebook page'
+PROJECT = 'project'
+COMMENTS = 'comments'
+ASSAY_TREE_RULE = 'assay tree rule'
+PROTOCOL_STATUS = 'protocol status'
+ASSAY_STAGE = 'assay stage'
+SCIENTIST = 'scientist'
+ASSAY_PRINCIPLE = 'assay principle'
+CREATION_DATE = 'creation date'
+PROTOCOL_DETAILS = 'protocol details'
+NOTEBOOK = 'notebook'
+
+
+class Protocol(SimpleLsThing):
```

Comment:
It seems weird / wrong to me to have Protocol inherit from SimpleLsThing. I can see that we want to use the same pythonic dictionary representation of an AbstractThing object, but there is LsThing-specific stuff such as `first_ls_things` / `second_ls_things` in the SimpleLsThing class.

I know it's more work, but I'd prefer to see a refactor that creates an AbstractSimpleThing class to have the common logic between this Protocol class and true LsThing-table-based SimpleLsThings

Reply:
Yeah, this mostly makes sense to me.  A few notes:

- Protocols and experiments do actually have interactions with itx states and values but are unused for years
- We should decide where exactly we want the abstractions to go because right now ls_states and ls_labels ...etc. live on the LsThing class so things like "get_preferred_label" are currently on the LsThing class https://github.com/mcneilco/acasclient/blob/1b638bfae9a78a4d98c1537164ada3d0ad17e7f4/acasclient/lsthing.py#L1170

Reply:
We do already have `AbstractThing`, `AbstractLabel`, `AbstractState` and `AbstractValue`.

I think we could move the `get_preferred_label` and the ls_states, ls_labels handling within `as_dict` up into AbstractThing.


I have some clear ideas about how we could refactor SimpleLsThing also. Let's separate this refactor I'm proposing from this PR. I can open up a PR for it.

Reply:
Ok yeah makes sense.  May end up being related but I started to pull additional fields up to the base model in this branch since they were shared fields.  Was also trying to simplify some of the initialization methods: https://github.com/mcneilco/acasclient/compare/DISCO-1604...DISCO-1604-protocol-refactor